core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -612,23 +616,17 @@ class RemoteLogManager(fetchLog: TopicPartition => Option[Log],
     val offset = fetchInfo.fetchOffset
     val maxBytes = Math.min(fetchMaxBytes, fetchInfo.maxBytes)
 
-    // get the epoch for the requested  offset from local leader epoch cache
-    // FIXME(@kamal), use the epochForOffset API instead of latest epoch.
-    //  val epoch = fetchLog(tp).map(log => log.leaderEpochCache.map(cache => cache.epochForOffset()))
-    var rlsMetadata: Optional[RemoteLogSegmentMetadata] = Optional.empty()
-    fetchLog(tp).foreach { log =>
-      log.leaderEpochCache.foreach(cache => {
-        var epoch = cache.latestEpoch
-        while (!rlsMetadata.isPresent && epoch.isDefined) {
-          rlsMetadata = fetchRemoteLogSegmentMetadata(tp, epoch.get, offset)
-          epoch = cache.findPreviousEpoch(epoch.get)
-        }
-      })
+    val epoch = fetchLog(tp).flatMap(log => log.leaderEpochCache.flatMap(cache => cache.epochForOffset(offset)))
+    val rlsMetadata = if (epoch.isDefined) {
+      fetchRemoteLogSegmentMetadata(tp, epoch.get, offset)
+    } else {
+      Optional.empty()
     }
 
     if (!rlsMetadata.isPresent) {
+      val epochStr = if (epoch.isDefined) epoch.get.toString else "NOT AVAILABLE"
```

Comment:
Nit (I'm really not sure if this is better though):

```
val epochStr = epoch.map(epoch => epoch.toString).getOrElse("NOT AVAILABLE")
```

Reply:
+1

Reply:
Thanks for the suggestion. Using if/else with `Option` in scala is considered as anti-practice (aka anti-functional programming), hence, your suggestion completely makes sense. Fixed.